File: packages/runar-py/runar/compile_check.py
"""Runar compile_check — validates contracts through the Python Rúnar frontend.

Mirrors the Go / Rust / Zig pattern: parse → validate → typecheck.
"""

import os


def compile_check(source_or_path: str, file_name: str | None = None) -> None:
    """Run Rúnar frontend (parse → validate → typecheck) on a contract source.

    Args:
        source_or_path: Either a file path to a .runar.{py,ts,sol,move,go,rs,rb,zig}
            file, or the source code string.
        file_name: Optional file name for error messages when passing source code
            directly. Must carry the `.runar.*` extension so the parser dispatcher
            picks the right frontend.

    Raises:
        RuntimeError: If any frontend pass reports errors.
    """
    if "\n" not in source_or_path and os.path.isfile(source_or_path):
        with open(source_or_path) as f:
            source = f.read()
        file_name = file_name or source_or_path
    else:
        source = source_or_path
        file_name = file_name or "contract.runar.py"

    try:
        from runar_compiler.frontend.parser_dispatch import parse_source
        from runar_compiler.frontend.validator import validate
        from runar_compiler.frontend.typecheck import type_check
    except ImportError as exc:
        raise RuntimeError(
            f"runar_compiler package not available — install the Python compiler "
            f"(cd compilers/python && pip install -e .): {exc}"
        ) from exc

    parse_result = parse_source(source, file_name)
    if parse_result.errors:
        raise RuntimeError(
            f"Parse errors in {file_name}:\n  "
            + "\n  ".join(parse_result.error_strings())
        )
    if parse_result.contract is None:
        raise RuntimeError(f"Parse returned no contract for {file_name}")

    validation = validate(parse_result.contract)
    if validation.errors:
        raise RuntimeError(
            f"Validation errors in {file_name}:\n  "
            + "\n  ".join(validation.error_strings())
        )

    typecheck = type_check(parse_result.contract)
    if typecheck.errors:
        raise RuntimeError(
            f"Type-check errors in {file_name}:\n  "
            + "\n  ".join(typecheck.error_strings())
        )

File: packages/runar-java/src/main/java/runar/lang/sdk/CompileCheck.java
package runar.lang.sdk;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import runar.compiler.frontend.ParserDispatch;
import runar.compiler.ir.ast.ContractNode;
import runar.compiler.passes.ExpandFixedArrays;
import runar.compiler.passes.Typecheck;
import runar.compiler.passes.Validate;

/**
 * Runs the Rúnar frontend (parse → validate → expand-fixed-arrays → typecheck)
 * on a Rúnar contract. Mirrors the Go {@code runar.CompileCheck},
 * Rust {@code runar::compile_check}, and Python {@code runar.compile_check}
 * APIs — frontend only, no codegen.
 *
 * <p>Use in JUnit tests alongside business-logic tests to ensure a contract
 * will compile to Bitcoin Script:
 *
 * <pre>{@code
 * @Test
 * void testCompile() throws Exception {
 *     CompileCheck.run(Path.of("Counter.runar.java"));
 * }
 * }</pre>
 */
public final class CompileCheck {

    private CompileCheck() {}

    /**
     * Run the Rúnar frontend on a contract source string. Dispatches by
     * the file extension carried in {@code fileName} (must end in one of
     * {@code .runar.{ts,sol,move,go,rs,py,zig,rb,java}}).
     *
     * @throws CompileException if any frontend pass reports errors
     */
    public static void check(String source, String fileName) {
        ContractNode contract;
        try {
            contract = ParserDispatch.parse(source, fileName);
        } catch (ParserDispatch.ParseException e) {
            throw new CompileException(
                "parse errors in " + fileName,
                List.of(e.getMessage()),
                e
            );
        } catch (RuntimeException e) {
            throw new CompileException(
                "parse errors in " + fileName,
                List.of(e.getMessage() == null ? e.toString() : e.getMessage()),
                e
            );
        }

        try {
            Validate.run(contract);
        } catch (Validate.ValidationException e) {
            throw new CompileException(
                "validation errors in " + fileName,
                e.errors(),
                e
            );
        }

        try {
            contract = ExpandFixedArrays.run(contract);
        } catch (ExpandFixedArrays.ExpandException e) {
            throw new CompileException(
                "expand-fixed-arrays errors in " + fileName,
                e.errors(),
                e
            );
        }

        try {
            Typecheck.run(contract);
        } catch (Typecheck.TypeCheckException e) {
            throw new CompileException(
                "type-check errors in " + fileName,
                e.errors(),
                e
            );
        }
    }

    /**
     * Read a Rúnar contract from disk and run the frontend on it.
     * Equivalent to {@code check(Files.readString(file), file.toString())}.
     *
     * @throws IOException      if the file cannot be read
     * @throws CompileException if any frontend pass reports errors
     */
    public static void run(Path file) throws IOException {
        String source = Files.readString(file);
        check(source, file.toString());
    }
}
